tools/ci/check_c_api_usage.py

```diff
@@ -0,0 +1,211 @@
+    extensions in 'exts' (e.g., '.c.src', '.c', '.h').
+    Excludes directories whose names are in 'excludes'.
+    """
+    results: list[Path] = []
+
+    for dirpath, dirnames, filenames in os.walk(root):
+        # Prune excluded directories
+        dirnames[:] = [d for d in dirnames if d not in excludes]
+        for fn in filenames:
+            # endswith handles mult-suffice patterns, e.g., .c.src
+            if any(fn.endswith(ext) for ext in exts):
+                results.append(Path(dirpath) / fn)
+    return results
+
+
+def scan_file(path: Path) -> list[tuple[str, int, str, str]]:
+    """
+    Scan a single file.
+    Returns list of (func_name, line_number, path_str, raw_line_str).
+    """
+    hits: list[tuple[str, int, str, str]] = []
+    in_block = False
+
+    try:
+        with path.open("r", encoding="utf-8", errors="ignore") as f:
+            for lineno, raw in enumerate(f, 1):
+                # Skip if approved by noqa markers
+                if NOQA_OK in raw or NOQA_MANUAL in raw:
+                    continue
+
+                # Remove comments; if nothing remains, skip
+                code, in_block = strip_comments(raw.rstrip("\n"), in_block)
+                if not code.strip():
+                    continue
+
+                # Find all suspicious calls in non-comment code
+                for m in FUNC_RX.finditer(code):
+                    func = m.group(0)
+                    hits.append((func, lineno, str(path), raw.rstrip("\n")))
+    except FileNotFoundError:
+        # File may have disappeared; ignore gracefully
+        pass
+    return hits
+
+
+def main(argv: list[str] | None = None) -> int:
+    ap = argparse.ArgumentParser(description="Borrow-ref C API linter (Python).")
+    ap.add_argument(
+        "--root",
+        default="numpy",
+        type=str,
+        help="Root directory to scan (default: numpy)"
+        )
+    ap.add_argument(
+        "--ext",
+        action="append",
+        default=None,
+        help="File extension(s) to include (repeatable). Defaults to .c,.h,.c.src,.cpp",
+    )
+    ap.add_argument(
+        "--exclude",
+        action="append",
+        default=None,
+        help="Directory name(s) to exclude (repeatable). Default: pythoncapi-compat",
+    )
+    args = ap.parse_args(argv)
+
+    root = Path(args.root)
+    if not root.exists():
+        print(f"error: root '{root}' does not exist", file=sys.stderr)
+        return 2
+
+    exts = set(args.ext) if args.ext else set(DEFAULT_EXTS)
+    excludes = set(args.exclude) if args.exclude else set(DEFAULT_EXCLUDES)
+
+    files = list(iter_source_files(root, exts, excludes))
+    print(f"Scanning {len(files)} C/C++ source files...")
+
+    # Output file (mirrors your shell behavior)
+    tmpdir = Path(".tmp")
+    tmpdir.mkdir(exist_ok=True)
+    fd, outpath = tempfile.mkstemp(
+        prefix="c_api_usage_report.",
+        suffix=".txt",
+        dir=tmpdir
+        )
+    os.close(fd)
```

Comment:
I suppose we could try to use a context manager to avoid the manual resource handling